lib/fetch/index.js

```diff
@@ -139,116 +185,146 @@ function fetch (input, init = undefined) {
+      // 1. If locallyAborted is true, terminate these substeps.
+      if (locallyAborted) {
+        return Promise.resolve()
+      }
 
-      abortFetch(p, request, responseObject, controller.serializedAbortReason)
-      return Promise.resolve()
-    }
+      try {
+        // 2. If response’s aborted flag is set, then:
+        if (response.aborted) {
+          // 1. Let deserializedError be the result of deserialize a serialized
+          //    abort reason given controller’s serialized abort reason and
+          //    relevantRealm.
 
-    // 3. If response is a network error, then reject p with a TypeError
-    // and terminate these substeps.
-    if (response.type === 'error') {
-      p.reject(new TypeError('fetch failed', { cause: response.error }))
-      return Promise.resolve()
-    }
+          // 2. Abort the fetch() call with p, request, responseObject, and
+          //    deserializedError.
 
-    // 4. Set responseObject to the result of creating a Response object,
-    // given response, "immutable", and relevantRealm.
-    responseObject = fromInnerResponse(response, 'immutable', relevantRealm)
+          abortFetch(p, request, responseObject, controller.serializedAbortReason, context)
+          return Promise.resolve()
+        }
 
-    // 5. Resolve p with responseObject.
-    p.resolve(responseObject)
-  }
+        // 3. If response is a network error, then reject p with a TypeError
+        // and terminate these substeps.
+        if (response.type === 'error') {
+          publishIfSubscribed('fetchError', context, response.error)
+          p.reject(new TypeError('fetch failed', { cause: response.error }))
+          return Promise.resolve()
+        }
 
-  controller = fetching({
-    request,
-    processResponseEndOfBody: handleFetchDone,
-    processResponse,
-    dispatcher: init?.dispatcher ?? getGlobalDispatcher() // undici
-  })
+        // 4. Set responseObject to the result of creating a Response object,
+        // given response, "immutable", and relevantRealm.
+        responseObject = fromInnerResponse(response, 'immutable', relevantRealm)
+
+        if (subscribers.asyncStart || subscribers.asyncEnd) {
+          context.result = responseObject[kState]
+        }
+
+        // 5. Resolve p with responseObject.
+      } finally {
+        publishIfSubscribed('asyncStart', context)
+        publishIfSubscribed('asyncEnd', context)
+        p.resolve(responseObject)
+      }
+    }
 
+    controller = fetching({
+      request,
+      processResponseEndOfBody: handleFetchDone,
+      processResponse,
+      dispatcher: init?.dispatcher ?? getGlobalDispatcher() // undici
+    })
+  } finally {
+    publishIfSubscribed('end', context)
+  }
   // 14. Return p.
+  if (subscribers.asyncStart || subscribers.asyncEnd) {
+    return p.promise.then(() => {
+      publishIfSubscribed('asyncStart', context)
+      publishIfSubscribed('asyncEnd', context)
+      context.result = null
+    })
```

Comment:
here the second bug:
```suggestion
  if (subscribers.asyncStart || subscribers.asyncEnd) {
    return p.promise.then((result) => {
      publishIfSubscribed('asyncStart', context)
      publishIfSubscribed('asyncEnd', context)
      context.result = null
      return result
    })
```